compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -258,251 +258,168 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
+                if let Some(new_required_bounds) = &mut new_required_bounds {
+                    new_required_bounds.retain(|b| item_required_bounds.contains(b));
+                } else {
+                    new_required_bounds = Some(item_required_bounds);
                 }
             }
-        }
 
-        // Imagine we have:
-        // ```
-        // trait Foo {
-        //   type Bar<'me>;
-        //   fn gimme(&self) -> Self::Bar<'_>;
-        //   fn gimme_default(&self) -> Self::Bar<'static>;
-        // }
-        // ```
-        // We only want to require clauses on `Bar` that we can prove from *all* functions (in this
-        // case, `'me` can be `static` from `gimme_default`)
-        match clauses.as_mut() {
-            Some(clauses) => {
-                clauses.drain_filter(|p| !function_clauses.contains(p));
-            }
-            None => {
-                clauses = Some(function_clauses);
+            if let Some(new_required_bounds) = new_required_bounds {
+                let required_bounds = required_bounds_by_item.entry(gat_def_id).or_default();
+                if new_required_bounds != *required_bounds {
+                    *required_bounds = new_required_bounds;
+                    // Iterate until our required_bounds no longer change
+                    // Since they changed here, we should continue the loop
+                    should_continue = true;
+                }
             }
         }
+        if !should_continue {
+            break;
+        }
     }
 
-    // If there are any clauses that aren't provable, emit an error
-    let clauses = clauses.unwrap_or_default();
-    debug!(?clauses);
-    if !clauses.is_empty() {
-        let param_env = tcx.param_env(trait_item.def_id);
+    for (gat_def_id, required_bounds) in required_bounds_by_item {
+        let gat_item_hir = tcx.hir().expect_trait_item(gat_def_id);
 
-        let mut clauses: Vec<_> = clauses
+        let gat_item = tcx.associated_item(gat_def_id);
+        let gat_generics = tcx.generics_of(gat_def_id);
+        if gat_item.kind != ty::AssocKind::Type || gat_generics.params.is_empty() {
+            continue;
+        }
+
+        debug!(?required_bounds);
+        let param_env = tcx.param_env(gat_def_id);
+        let gat_hir = gat_item_hir.hir_id();
+
+        let mut unsatisfied_bounds: Vec<_> = required_bounds
             .into_iter()
             .filter(|clause| match clause.kind().skip_binder() {
                 ty::PredicateKind::RegionOutlives(ty::OutlivesPredicate(a, b)) => {
-                    !region_known_to_outlive(
-                        tcx,
-                        trait_item.hir_id(),
-                        param_env,
-                        &FxHashSet::default(),
-                        a,
-                        b,
-                    )
+                    !region_known_to_outlive(tcx, gat_hir, param_env, &FxHashSet::default(), a, b)
```

Comment:
Diff might be better here to split this into two parts? (Or two commits)

Reply:
Good point. I'll make this diff history a bit cleaner. Just wanted to wait to make sure the algorithm I implemented wasn't totally off the mark, but I can make things a bit neater now I'm not completely wrong :)